src/nodes.coffee

```diff
@@ -1154,19 +1154,13 @@ exports.Value = class Value extends Base
     ret = @base.ast()
     for prop, propIndex in @properties
       ret =
-        mergeAstLocationData(
-          Object.assign {
-            type: 'MemberExpression'
-            object: ret
-            property: prop.ast()
-            computed: prop instanceof Index or prop.name?.unwrap() not instanceof PropertyName
-            optional: !!prop.soak
-            shorthand: !!prop.shorthand
-          }, prop.astLocationData()
-          ret
-        )
-      if propIndex is 0 and @base instanceof Parens and @base.locationData?
-        mergeAstLocationData ret, @base.astLocationData()
+        type: 'MemberExpression'
+        object: ret
+        property: prop.ast()
+        computed: prop instanceof Index or prop.name?.unwrap() not instanceof PropertyName
+        optional: !!prop.soak
+        shorthand: !!prop.shorthand
+      Object.assign ret, mergeAstLocationData(@base.astLocationData(), prop.astLocationData())
```

Comment:
I think this looks correct, including handling the "special case" when the `base` is wrapped in parens. Just thinking through it:
- For the simple case where there are no `properties`, the location data should correctly not include the parens since `@base.ast()` will call `Parens::ast()`, which forwards to its `@body.unwrap().ast()`
- When there are `properties`, this looks right as we want to span the `base`'s location data through that "currently outermost" `prop`'s location data
- For the special case with parens around the `base`, using `@base.astLocationData()` seems like an elegant way to handle that since `@base.astLocationData()` will call `Parens::astLocationData()` (ie `Base::astLocationData()`, which should use the `Parens`' `locationData`)

At this point you could include the merged location data in the assignment to `ret`:
```
for prop, propIndex in @properties
  ret = Object.assign {
    type: 'MemberExpression'
    ...
  }, mergeAstLocationData(@base.astLocationData(), prop.astLocationData())
```

So in this instance, the combination of `.astLocationData()` and the immutable version of `mergeAstLocationData()` seems to have resulted in cleaner code 👍

Reply:
Yes. And your explanation would be good to add as a comment.

And sure, we could do this all in an `Object.assign` one-liner, but does that really make it any clearer? I like the separation of one line doing the node properties and another line doing the location data. I find it easier to follow the more separation we have between the three groups of AST properties (type, “properties” and location data).

That’s what I wanted to discuss in a follow-up PR anyway, so maybe we can merge this in and then look at the organization of how we’re assembling these AST nodes?

Reply:
@GeoffreyBooth sure, added a version of the explanation as a comment and merged